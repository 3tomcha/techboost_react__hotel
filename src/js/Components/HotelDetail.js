import React, {Fragment} from 'react';
import ReactDOM from 'react-dom';
import RakutenTravelApi from '../libs/RakutenTravelApi.js';
import queryString from 'query-string';
import {Card, Row, Col, Button} from 'react-bootstrap';
import Header from './Header';
import {FaStar, FaRegStar, FaYenSign, FaCommentDots} from 'react-icons/fa';
import { setHoteldetail } from '../redux/actions';
import { connect } from 'react-redux';


class HotelDetail extends React.Component {
  constructor(props){
    super(props);
    this.id = this.props.match.params.id;
    this.checkInDay = queryString.parse(this.props.location.search).checkInDate;
    this.checkOutDay = queryString.parse(this.props.location.search).checkOutDate;
  }

  async componentDidMount(){
      // checkInDay,checkOutDayを元に、楽天APIを用いてホテル詳細情報を取得する
    const response = await RakutenTravelApi.fetchDetailPlan(
      this.id,
      this.checkInDay,
      this.checkOutDay
    );

    const hotel = JSON.parse(response).hotels[0].hotel;
    const basicInfo = hotel[0].hotelBasicInfo;
    const roomInfo = hotel[0].roomInfo;

    this.props.setHoteldetail({
      hotelName: basicInfo.hotelName,
      reviewAverage: basicInfo.reviewAverage,
      hotelImageUrl: basicInfo.hotelImageUrl,
      hotelSpecial: basicInfo.hotelSpecial,
      access: basicInfo.access,
      userReview: basicInfo.userReview,
      roomInfo: roomInfo
    });
  }

  render() {
    // console.log(this.state);
    console.log(this.props);
    // ユーザー評価によって星の数を増減
    const stars = [...Array(5).keys()].map(i => {
      return (i < parseInt(this.props.hotel.reviewAverage))? <FaStar key={i} /> : <FaRegStar key={i} />;
  });
  return(
    <div className="hoteldetail pagecontainer">
      <Header/>
      <Card>
        <Card.Body>
          <Row>
            <Col md={3}>
              <Card.Img variant="top" src={this.props.hotel.hotelImageUrl} />
            </Col>
            <Col md={7}>
              <Card.Title>
                {this.props.hotel.hotelName}
              </Card.Title>
              <Card.Text>
                {this.props.hotel.hotelSpecial}
              </Card.Text>
              <Card.Text>
                {this.props.hotel.access}
              </Card.Text>
              <Card.Text>
                {stars}
                {this.props.hotel.reviewAverage}
              </Card.Text>
              <Card.Text>
                {this.props.hotel.userReview}
                <FaCommentDots/>
              </Card.Text>
              <Button variant="primary"
                onClick={
                  () => {
                    this.props.history.goBack();
                  }
                }>戻る</Button>
              </Col>
            </Row>
          </Card.Body>
        </Card>
      </div>
    );
  }
}

// storeから得たstateは、propsのhotelに変更
function mapStateToProps(state){
  return { hotel: state.hoteldetail};
}

export default connect(
  mapStateToProps,
  {setHoteldetail}
)(HotelDetail);
