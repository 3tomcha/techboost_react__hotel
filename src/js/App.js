import React from 'react';
import ReactDOM from 'react-dom';
import Home from './Home';
import SearchForm from './Components/SearchForm';
import HotelList from './Components/HotelList';
import HotelDetail from './Components/HotelDetail';
import { BrowserRouter, Route, Switch, Link } from 'react-router-dom'

function App(props){
  const today = new Date();
  const tomorrow = new Date(
    today.getFullYear(),
    today.getMonth(),
    today.getDate() + 1,
    today.getHours(),
    today.getMinutes()
  );

  return(
    <BrowserRouter>
      <Switch>
        <Route path="/hotels/:id" component={HotelDetail}/>
        <Route path="/" component={Home}/>
        <Route path="/hotellist" component={HotelList}/>
      </Switch>
    </BrowserRouter>
  );
}
export default App;
