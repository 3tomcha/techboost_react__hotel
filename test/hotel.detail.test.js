import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import HotelDetail from '../src/js/Components/HotelDetail.js';
import App from '../src/js/App.js';
import renderer from 'react-test-renderer';
import { Link, Route } from "react-router-dom"
import {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils';
Enzyme.configure({ adapter: new Adapter() });

test("detail", () => {
  let wrapper = renderer(
    <MemoryRouter initialEntries={["/hotels/141356?checkInDate=2019-07-12&checkOutDate=2019-07-13"]}>
    <MainRouter />
    </MemoryRouter>

    // <Link to="/hotels/141356?checkInDate=2019-07-12&checkOutDate=2019-07-13">
      // <a href="/hotels/141356?checkInDate=2019-07-12&checkOutDate=2019-07-13">
      //   <button>詳細ボタン</button>
      // </a>
    // </Link>
  );
  // wrapper.find('button').simulate('click');
  console.log(wrapper.debug());


  // console.log(link);
  // console.log(link.debug());

  // const component = renderer.create(
  //   <MemoryRouter initialEntries=
  //   {['hotels/141356?checkInDate=2019-07-15&checkOutDate=2019-07-16']}>
  //   <App />
  //   </MemoryRouter>
  //
  //   // <Route component={props => <HotelDetail {...props} />}
  //   // path='hotels/141356?checkInDate=2019-07-15&checkOutDate=2019-07-16' />
  //     // <Route path="/hotels/:id" component={HotelDetail}></Route>
  //   // <MemoryRouter initialEntries=
  //   // {['hotels/141356?checkInDate=2019-07-15&checkOutDate=2019-07-16']}>
  //   // <Route component={props => <HotelDetail {...props} />}
  //   // path='hotels/141356?checkInDate=2019-07-15&checkOutDate=2019-07-16' />
  //   // </MemoryRouter>
  // );
  // let tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});
