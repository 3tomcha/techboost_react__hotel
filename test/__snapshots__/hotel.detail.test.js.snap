// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail1 1`] = `
<div>
  HotelDetail 
  141356
  <ul>
    <li>
      
    </li>
    <li>
      
    </li>
    <li>
      
    </li>
    <li>
      
    </li>
    <li>
      
    </li>
    <li>
      
    </li>
    // 
    <li />
    // 
    <li />
    // 
    <li />
  </ul>
</div>
`;
